### DISCUSSION TITLE
Visualization data process problem -- Folium

### DISCUSSION THREAD

**Discussion by Davidyang1989** [datapane/datapane#107] [Help]
I tried to add the Search plugin in the codes following, but it always feeds back the KeyError as follow after the codes.
I dont know where i missed the value?
This is the link without the desired search function (https://datapane.com/u/han/reports/clients-map/)
Thanks for your help!

============================
import folium
import datapane as dp 
import pandas as pd

data = pd.read_excel('data.xlsx')
data.dropna(axis=0,inplace=True,how='any')
data1= data[4:250]

m = folium.Map(location=[35.0233,107.8377], zoom_start=4, tiles="OpenStreetMap")

tooltip = "blabla"

folium.Marker(
    location=[25.0407974484392, 121.573737859928],
    popup = folium.Popup(html="简称：新光人寿 <br/>销售王五 <br/>手机：13999999999<br/>对接开始时间：20220101<br/> \
    台湾省台北市忠孝西路一段66号<br/>其他补充信息（做展示用）",
                         parse_html=False, max_width=200, show=False, sticky=False),    
    icon=folium.Icon(color="darkblue", icon="info-sign"),  
    fill=False,
    draggable=False,
    tooltip=tooltip
).add_to(m) 

folium.Marker(
    location=[39.9134370547403, 116.453599397516],
    popup=folium.Popup('新华人寿<br/>李四<br/>20230501<br/>其他信息', parse_html=False),
    icon=folium.Icon(color="green", icon="info-sign"),
    fill=False,
    draggable=False,
    tooltip=tooltip
    
).add_to(m) # 

def automation(Latitude,Longtitude,shortName):
    folium.Marker(
        location=[Latitude, Longtitude],
        popup="%s<br/>销售X<br/>yyyymmdd<br/>其他信息"%(shortName),
        icon=folium.Icon(color="darkblue", icon="users"),  
        draggable=False,
        tooltip=  "点击显示机构名称、销售和开始负责本机构的时间"   
    ).add_to(m) 
    
for x,y,z in zip(data1.Latitude,data1.Longtitude,data1.shortName):
    #print(x,y,z)
    automation(x,y,z)

from folium.plugins import Search

# Add search 

tempoTable = data1[['name', 'types', 'shortName','address','geometry']]

from collections import defaultdict
dataJson = defaultdict(dict)
print(dataJson)
for name,types,shortName,address,geometry in tempoTable.itertuples(index=False,name='shortName'):
    #print(name)
    dataJson[shortName] = name,types,shortName,address,geometry

#dataJson.pop('features') #删除features（不知从何处来的这个key和对应空值values）
#print(dataJson)

citysearch = Search(
    layer=folium.GeoJson(dataJson),
    search_label='shortName',
    geom_type="Point",
    placeholder="Search for a Clients"
).add_to(m)    

m.save("index.html")

====================================

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-14-7e54a76380dd> in <module>
     96 
     97 print("ok")
---> 98 m.save("index.html")
     99 
    100 # Sign-in with your unique token

~\Anaconda3\lib\site-packages\branca\element.py in save(self, outfile, close_file, **kwargs)
    165 
    166         root = self.get_root()
--> 167         html = root.render(**kwargs)
    168         fid.write(html.encode('utf8'))
    169         if close_file:

~\Anaconda3\lib\site-packages\branca\element.py in render(self, **kwargs)
    317         """Renders the HTML representation of the element."""
    318         for name, child in self._children.items():
--> 319             child.render(**kwargs)
    320         return self._template.render(this=self, kwargs=kwargs)
    321 

~\Anaconda3\lib\site-packages\folium\folium.py in render(self, **kwargs)
    368             '</style>'), name='map_style')
    369 
--> 370         super(Map, self).render(**kwargs)
    371 
    372     def fit_bounds(self, bounds, padding_top_left=None,

~\Anaconda3\lib\site-packages\branca\element.py in render(self, **kwargs)
    641 
    642         for name, element in self._children.items():
--> 643             element.render(**kwargs)

~\Anaconda3\lib\site-packages\folium\plugins\search.py in render(self, **kwargs)
    113     def render(self, **kwargs):
    114         if isinstance(self.layer, GeoJson):
--> 115             keys = tuple(self.layer.data['features'][0]['properties'].keys())
    116         elif isinstance(self.layer, TopoJson):
    117             obj_name = self.layer.object_path.split('.')[-1]

KeyError: 0

**Comment by johnmicahreid** [comment]
Hi David, it looks like the Folium error is coming from the folium lib rather than datapane- the error happens when you call folium.Map.save. Maybe you are missing some data? See this discussion thread: https://www.kaggle.com/questions-and-answers/42673
